libraries/AP_IBus_Telem/AP_IBus_Telem.cpp

```diff
@@ -0,0 +1,193 @@
+*/
+
+#include <AP_IBus_Telem/AP_IBus_Telem.h>
+#include <AP_BattMonitor/AP_BattMonitor.h>
+#include <AP_Vehicle/AP_Vehicle.h>
+#include <AP_Arming/AP_Arming.h>
+
+#if HAL_IBUS_TELEM_ENABLED
+
+void AP_IBus_Telem::init()
+{
+    const AP_SerialManager &serial_manager = AP::serialmanager();
+    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_IBUS_Telem, 0))) {
+        _port->set_options(_port->OPTION_HDPLEX);
+        _port->begin(AP_SERIALMANAGER_IBUS_TELEM_BAUD);
+        _initialized = true;
+    }
+}
+
+void AP_IBus_Telem::populate_checksum(uint8_t *packet, uint16_t size)
+{
+    uint16_t checksum = 0xFFFF;
+
+    for (int i=0; i<size-2; i++) {
+        checksum -= *packet;
+        packet++;
+    }
+
+    *packet = checksum & 0x0ff;
+    packet++;
+    *packet = checksum >> 8;
+}
+
+void AP_IBus_Telem::loop(void)
+{
+    //no need to run the loop if not initialized or no sensors added
+    if (!_initialized) {
+        return;
+    }
+
+    uint32_t now = AP_HAL::millis();
+    if (now - _last_send_measurements_time > UPDATE_MEASUREMENTS_TIMING) {
+        setSensorMeasurements();
+    }
+
+    // only process data already in our UART receive buffer
+    while (_port->available() > 0) {
+        // only consider a new data package if we have not heard anything for >3ms
+        if (now - _last >= PROTOCOL_TIMEGAP) {
+            _state = GET_LENGTH;
+        }
+        _last = now;
+
+        int16_t c = _port->read();
+        if (c < 0) {
+            return;
+        }
+
+        switch (_state) {
+        case GET_LENGTH:
+            if (c <= PROTOCOL_MAX_LENGTH && c > PROTOCOL_OVERHEAD) {
+                _ptr = 0;
+                _len = c - PROTOCOL_OVERHEAD;
+                _chksum = 0xFFFF - c;
+                _state = GET_DATA;
+            } else {
+                _state = DISCARD;
+            }
+            break;
+
+        case GET_DATA:
+            _buffer[_ptr++] = c;
+            _chksum -= c;
+            if (_ptr == _len) {
+                _state = GET_CHKSUML;
+            }
+            break;
+
+        case GET_CHKSUML:
+            _lchksum = c;
+            _state = GET_CHKSUMH;
+            break;
+
+        case GET_CHKSUMH:
+            // Validate checksum
+            if (_chksum == (c << 8) + _lchksum) {
+                // Checksum is all fine Execute command -
+                uint8_t adr = _buffer[0] & 0x0f;
+                if (adr > sizeof(_sensors) || adr <=0 || _len != 1) {
+                    return;
+                }
+
+                // all sensor data commands go here
+                // we only process the len==1 commands (=message length is 4 bytes incl overhead) to prevent the case the
+                // return messages from the UART TX port loop back to the RX port and are processed again. This is extra
+                // precaution as it will also be prevented by the PROTOCOL_TIMEGAP required
+                _sensorinfo *s = &_sensors[adr - 1];
```

Comment:
This is an out-of-bounds-read depending on what the user has sent us.  That's not good.

Need to check against the length of the sensors array.